app/V1Module/presenters/AssignmentSolutionReviewsPresenter.php

```diff
@@ -0,0 +1,407 @@
+     * @Param(type="post", name="file", validation="string:1..256", required=true,
+     *        description="Identification of the file to which the comment is related to.")
+     * @Param(type="post", name="line", validation="numericint", required=true,
+     *        description="Line in the designated file to which the comment is related to.")
+     * @Param(type="post", name="issue", validation="bool", required=false,
+     *        description="Whether the comment is an issue (expected to be resolved by the student)")
+     * @Param(type="post", name="suppressNotification", validation="bool", required=false,
+     *        description="If true, no email notification will be sent (only applies when the review has been closed)")
+     * @param string $id identifier of the solution
+     * @throws InternalServerException
+     */
+    public function actionNewComment(string $id)
+    {
+        $solution = $this->assignmentSolutions->findOrThrow($id);
+        if ($solution->getReviewStartedAt() === null) {
+            // first comment also opens the review
+            $solution->setReviewStartedAt(new DateTime());
+            $this->assignmentSolutions->persist($solution);
+        }
+
+        // get and verify inputs
+        $req = $this->getRequest();
+        $text = trim($req->getPost("text"));
+        if (!$text) {
+            throw new BadRequestException("The text of the comment must not be empty.");
+        }
+
+        $file = trim($req->getPost("file"));
+        $line = $req->getPost("line");
+        $this->verifyCodeLocation($solution, $file, $line);
+        $issue = filter_var($req->getPost("issue"), FILTER_VALIDATE_BOOLEAN);
+
+        // create the review comment
+        $comment = new ReviewComment($solution, $this->getCurrentUser(), $file, $line, $text, $issue);
+        $this->reviewComments->persist($comment);
+
+        if ($solution->getReviewedAt() !== null) {
+            // review is already closed, this needs special treatement
+            if ($issue) {
+                $solution->setIssuesCount($solution->getIssuesCount() + 1);
+                $this->assignmentSolutions->persist($solution);
+            }
+
+            $suppressNotification = filter_var($req->getPost("suppressNotification"), FILTER_VALIDATE_BOOLEAN);
+            if (!$suppressNotification) {
+                $this->reviewsEmailSender->newReviewComment($solution, $comment);
+            }
+        }
+
+        $this->sendSuccessResponse($comment);
+    }
+
+    public function checkEditComment(string $id, string $commentId)
+    {
+        $solution = $this->assignmentSolutions->findOrThrow($id);
+        $comment = $this->reviewComments->findOrThrow($commentId);
+        if (!$this->assignmentSolutionAcl->canEditReviewComment($solution, $comment)) {
+            throw new ForbiddenRequestException("You are not allowed to edit this review comment");
+        }
+
+        if ($comment->getSolution()->getId() !== $id) {
```

Comment:
Couldn't this be done in the permissions?

Reply:
Technically yes, but I did not like it. This is not a matter of ACL, this is a matter of verifying endpoint inputs. The endpoint gets two IDs, where one is actually implied by the other. I did not want to change the endpoint (for reasons risen at the frontend), so verification must be performed (also note that I am throwing BadRequest, not Forbidden exception).

Reply:
Ok, makes sense